firedrake/slate/slac/kernel_builder.py

```diff
@@ -580,14 +580,15 @@ def collect_coefficients(self):
         coeffs = self.expression.coefficients()
         coeff_dict = OrderedDict()
         for i, c in enumerate(coeffs):
-            element = c.ufl_element()
-            if type(element) == MixedElement:
-                mixed = OrderedDict()
-                for j, c_ in enumerate(c.split()):
+            coeff = c.orig_function if isinstance(c, slate.BlockFunction) else c
+            if type(coeff.ufl_element()) == MixedElement:
+                mixed = OrderedDict() if coeff not in coeff_dict else coeff_dict[coeff]
+                splits = c.split_function if isinstance(c, slate.BlockFunction) else coeff.split()
+                for j, c_ in enumerate(splits):
                     mixed[c_] = f"w_{i}_{j}", self.extent(c_)
-                coeff_dict[c] = mixed
+                coeff_dict.update({coeff: mixed})
```

Comment:
I think, here, we can loop over `self.expression.coeff_map` instead of over `self.expression.coefficients()`, since we have already removed duplication in `coeff_map`?

Reply:
The `coeff_map` does not carry enough information to generate the kernel arguments here. In particular we need the extent. We either need the extent as explicit information or we can deduce it from a coefficient (like we are doing atm)